File: Practicas/Qytetet/QytetetRuby/lib/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class TituloPropiedad
    attr_accessor :hipotecada, :propietario
    attr_reader :num_hoteles, :num_casas, :nombre, :precio_compra, :alquiler_base,
      :hipoteca_base, :precio_edificar, :factor_revalorizacion
    def initialize(hip, num_h, num_c, name, pr_com, al_b, hip_b, pr_ed, fac_rev)
      @hipotecada = hip
      @num_hoteles = num_h
      @num_casas = num_c
      @nombre = name
      @precio_compra = pr_com
      @alquiler_base = al_b
      @hipoteca_base = hip_b
      @precio_edificar = pr_ed
      @factor_revalorizacion = fac_rev
    end
    
    def self.inicializar_propiedad(name, pr_com, al_b, hip_b, pr_ed, fac_rev)
      new(false, 0, 0, name, pr_com, al_b, hip_b, pr_ed, fac_rev)
    end
    
    def calcular_coste_cancelar
      (calcular_coste_hipotecar * 1.1).to_i 
    end
    
    def calcular_coste_hipotecar
      (@hipoteca_base + @num_casas * 0.5 * @hipoteca_base + @num_hoteles * @hipoteca_base).to_i
    end
    
    def calcular_importe_alquiler
      @alquiler_base + ((@num_casas * 0.5 + @num_hoteles * 2) * @precio_edificar).to_i 
    end
    
    def calcular_precio_venta
      (@precio_compra + (@num_casas + @num_hoteles) * @precio_edificar * @factor_revalorizacion).to_i
    end
    
    def cancelar_hipoteca
      @hipotecada = false
    end
    
    def edificar_casa
      @num_casas = @num_casas + 1
    end
    
    def edificar_hotel
      @num_casas = @num_casas - 4
      @num_hoteles = @num_hoteles + 1 
    end
    
    def hipotecar
      coste_hipoteca = calcular_coste_hipotecar
      @hipotecada = true
      coste_hipoteca
    end
    
    def pagar_alquiler
      coste_alquiler = calcular_importe_alquiler
      @propietario.modificar_saldo(coste_alquiler)
      coste_alquiler
    end
    
    def propietario_encarcelado
      @propietario.encarcelado
    end
    
    def tengo_propietario
      @propietario != nil
    end
    
    def to_s
      "TituloPropiedad{nombre=#{@nombre}, hipotecada=#{@hipotecada}, " +
      "precio_compra=#{@precio_compra}, alquiler_base=#{@alquiler_base}, " +
      "hipoteca_base=#{@hipoteca_base}, precio_edificar=#{@precio_edificar}, " +
      "num_hoteles=#{@num_hoteles}, num_casas=#{@num_casas}, " +
      "factor_revalorizacion=#{@factor_revalorizacion}}"
    end
  private_class_method :new
  end
end

File: Practicas/Qytetet/QytetetJava/src/modeloqytetet/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author maxigang
 */
public class TituloPropiedad {
    
    private String nombre;
    private boolean hipotecada;
    private int precioCompra, alquilerBase, hipotecaBase, precioEdificar, numHoteles,
        numCasas;
    private double factorRevalorizacion;
    
    private Jugador propietario;
    
    public TituloPropiedad(String nombre, int precioCompra, int alquilerBase, 
            int hipotecaBase, int precioEdificar, double factorRevalorizacion){
        
        this.hipotecada = false;
        this.numHoteles = 0;
        this.numCasas = 0;
        
        this.nombre = nombre;
        this.precioCompra = precioCompra;
        this.alquilerBase = alquilerBase;
        this.hipotecaBase = hipotecaBase;
        this.precioEdificar = precioEdificar;
        this.factorRevalorizacion = factorRevalorizacion;
       
    }
    
    public TituloPropiedad(boolean hipotecada, int numHoteles, int numCasas,
            String nombre, int precioCompra, int alquilerBase, 
            int hipotecaBase, int precioEdificar, float factorRevalorizacion){
        
        this.hipotecada = hipotecada;
        this.numHoteles = numHoteles;
        this.numCasas = numCasas;
        
        this.nombre = nombre;
        this.precioCompra = precioCompra;
        this.alquilerBase = alquilerBase;
        this.hipotecaBase = hipotecaBase;
        this.precioEdificar = precioEdificar;
        this.factorRevalorizacion = factorRevalorizacion;
       
    }
    
    int calcularCosteCancelar(){
        return (int) (calcularCosteHipotecar() * 1.1);
    }
    
    int calcularCosteHipotecar(){
        return (int) (hipotecaBase + numCasas * 0.5 * hipotecaBase + numHoteles * hipotecaBase);
    }
    
    int calcularImporteAlquiler(){
        return alquilerBase + (int) ((numCasas*0.5 + numHoteles*2) * precioEdificar); // HE AÑADIDO LO DE PRECIOEDIFICAR!!!!!!!!!
    }
    
    int calcularPrecioVenta(){
        return (int) (precioCompra + (numCasas + numHoteles) * precioEdificar * factorRevalorizacion);
    }
    
    void cancelarHipoteca(){
        hipotecada = false;
    }
    
    void edificarCasa(){
        numCasas++;
    }
    
    void edificarHotel(){
        numCasas -= 4;
        numHoteles++;
    }
    
    public String getNombre() {
        return nombre;
    }

    public Boolean getHipotecada() {
        return hipotecada;
    }

    public int getPrecioCompra() {
        return precioCompra;
    }

    int getAlquilerBase() {
        return alquilerBase;
    }

    int getHipotecaBase() {
        return hipotecaBase;
    }

    int getPrecioEdificar() {
        return precioEdificar;
    }
    
    Jugador getPropietario() {
        return propietario;
    }

    int getNumHoteles() {
        return numHoteles;
    }

    int getNumCasas() {
        return numCasas;
    }

    double getFactorRevalorizacion() {
        return factorRevalorizacion;
    }
    
    int hipotecar(){
        int costeHipoteca = calcularCosteHipotecar();
        setHipotecada(true);
        return costeHipoteca;
    }
    
    int pagarAlquiler(){
        int costeAlquiler = calcularImporteAlquiler();
        propietario.modificarSaldo(costeAlquiler);
        return costeAlquiler;
    }
    
    boolean propietarioEncarcelado(){
        return propietario.getEncarcelado();
    }
    
    public void setHipotecada(Boolean hipotecada) {
        this.hipotecada = hipotecada;
    }
    
    void setPropietario(Jugador propietario){
        this.propietario = propietario;
    }
    
    boolean tengoPropietario(){
        return propietario != null;
    }
    @Override
    public String toString() {
        return "TituloPropiedad{" + "nombre=" + nombre + ", hipotecada=" + 
                hipotecada + ", precioCompra=" + precioCompra + 
                ", alquilerBase=" + alquilerBase + ", hipotecaBase=" 
                + hipotecaBase + ", precioEdificar=" + precioEdificar 
                + ", numHoteles=" + numHoteles + ", numCasas=" + numCasas + 
                ", factorRevalorizacion=" + factorRevalorizacion + '}';
    }
    
    
}
